### PR TITLE
Upgraded AWS API version to 2014-10-01 (latest)

### PR DESCRIPTION
**Pull request by dimitern**
- All EC2 requests now use the latest AWS API version available (2014-10-01).
- ec2test server response types changed as needed to mimic EC2 better (proper response type names and xmlns added, DescribeSecurityGroups response XML fixed to serialized multiple IPPerm.SourceIPs values).
- Changed ec2test responses and tests to match EC2 in a few cases (e.g. InvalidGroup.InUse -> DependencyViolation)
- Removed CreateSecurityGroupVPC() and changed CreateSecurityGroup() to take a VPC ID as first argument (can be empty to create an EC2-Classic SG or a valid VPC ID to create an EC2-VPC SG).
- When debug=true, requests and responses are dumped to the log output. Requests are formatted similarly to the way AWS API docs does).
- Added a few fields to the Instance type - AssociatePublicIP, EBSOptimized, SRIOVNetSupport - available in more recent AWS API versions.
- Improved tests logging and stability when running against live EC2 servers.

### CODE REVIEW

File: ec2/ec2.go [hunk 1]
```diff
@@ -787,15 +777,35 @@ type IPPerm struct {
...
+// SourceIP represents a single security group source IP address
+// within an ingress permission rule.
+//
+// See http://goo.gl/4oTxv for more details.
+type SourceIP struct {
+	IP string `xml:"cidrIp"`
+}
+
+// SourceIPs is a helper for constructing []SourceIP from strings.
+func SourceIPs(ips ...string) []SourceIP {
```
Comment (niemeyer) [inline ec2/ec2.go]: This helper doesn't seem worth it. It's just a plain slice:

```
[]SourceIP{{one}, {two}}
```

File: ec2/ec2.go [hunk 2]
```diff
@@ -787,15 +777,35 @@ type IPPerm struct {
 	Protocol     string              `xml:"ipProtocol"`
 	FromPort     int                 `xml:"fromPort"`
 	ToPort       int                 `xml:"toPort"`
-	SourceIPs    []string            `xml:"ipRanges>item>cidrIp"`
```
Comment (niemeyer) [inline ec2/ec2.go:790]: I couldn't find a test change that exercises the reason why this is being changed (what you described in the pull request summary). Perhaps it's here and I'm just not seeing it?

File: ec2/ec2.go [hunk 3]
```diff
@@ -295,7 +298,10 @@ type Instance struct {
 	VirtType           string             `xml:"virtualizationType"`
 	Monitoring         string             `xml:"monitoring>state"`
 	AvailZone          string             `xml:"placement>availabilityZone"`
+	AssociatePublicIP  bool               `xml:"associatePublicIpAddress,omitempty"`
 	PlacementGroupName string             `xml:"placement>groupName"`
+	EbsOptimized       bool               `xml:"ebsOptimized,omitempty"`
```
Comment (niemeyer) [inline ec2/ec2.go]: These should all be "EBS", equivalent to "EC2".

**Comment by niemeyer** [comment]
Besides the points above:
1. You haven't referenced the change log from Amazon for the API version change. I won't bother you again about this, but I'd strongly suggest doing a review so that the team and the community can know what to expect.
2. I suspect the change to simply have a vpcId in one relevant method is likely making the API inconsistent. It's quite probable that such an id is needed in other methods too. This is just a FYI, though. I don't expect you'll fix all of those before merging this long standing change. If somebody does care about the package, I would suggest stepping back and thinking this through for the next major API version bump.

**Comment by niemeyer** [comment]
Given all the above, please feel free to refer to the maintaining team for when to merge this.

File: ec2/ec2.go (hunk 1 above)
Comment (dimitern) [inline ec2/ec2.go]: OK, I'll drop the helper.

File: ec2/ec2.go (hunk 2 above)
Comment (dimitern) [inline ec2/ec2.go:790]: If you have a look at ec2t_test.go - TestIPPerms there is a case where this is exercised - starting from line 491. We set 2 IPPerm rules - 1 with two source groups (both by Id and Name) and source IP 127.0.0.0/24, 1 with just a source IP 200.1.1.34/32. Then, when we list the g0 security group the response from Amazon contains 3 IPPerm rules: 1 for each source group (by Id or Name) and 1 with the two source IPs. The way IPPerm.SourceIPs were defined before - `xml:"ipRanges>item>cidrIp"`, given this XML:

``` xml
...
<ipPermissions>
  <item>
    ...
   <ipRanges>
     <item><cidrIp>127.0.0.0/24</cidrIp></item>
     <item><cidrIp>200.1.1.34/32</cidrIp></item>
  </ipRanges>
  ...
```

Rather than SourceIPs: []string{"127.0.0.0/24", "200.1.1.34/32"} you'd get only the first one, because the expected XML format is `<ipRanges><item><cidrIp>127.0.0.0/24</cidrIp><cidrIp>200.1.1.34/32</cidrIp></item></ipRanges>` (and the ec2test server in fact returns this, but the live Amazon servers return a response with the format described above).

The easiest way to verify this is to set debug=true in ec2.go and run go test -gocheck.v -gocheck.f TestIPPerms on launchpad.net/goamz/ec2 - with and without the -amazon flag - the latter fails and you can see the response XML is different from Amazon and that response is not deserialized correctly.

File: ec2/ec2.go (hunk 3 above)
Comment (dimitern) [inline ec2/ec2.go]: OK, will change this.

File: ec2/ec2.go (hunk 2 above)
Comment (dimitern) [inline ec2/ec2.go:790]: Sorry, so a small correction. I tried what I suggested - running `go test -gocheck.v -gocheck.f TestIPPerms -amazon` on launchpad.net/goamz/ec2 and failed, but not due to the reason source IPs were not parsed correctly. It failed because:
1. The error code for calling AuthorizeSecurityGroupIngress with an invalid CIDR (z127.0.0.1/24) is InvalidParameterValue, not InvalidPermission.Malformed (even though the API documentation states the latter should be expected for invalid CIDR - http://docs.aws.amazon.com/AWSEC2/latest/APIReference/errors-overview.html)
2. The source IPs _get deserialized_ correctly - []string{"127.0.0.0/24", "200.1.1.34/32"} (well, almost - 127.0.0.1/24 from the test is converted to the equivalent 127.0.0.0/24 by Amazon), but the length of response.Groups[0].IPPerms is 3, not 1 as expected by the test.

As some of the changes in this PR modified the failing live tests to pass. As for the SourceIPs I experimented a bit - http://play.golang.org/p/mR31F_gOIG. It seems the encoding/xml package's Unmarshal function, for a struct field like: <pre>Zs []int `xml:"x>y>z"`</pre> parses both `<x><y><z>0</z><z>1</z></y></x>` and `<x><y><z>0</z></y><y><z>1</z></y></x>`  as []int{0,1} - slightly surprising to me for the latter case , but useful nevertheless and more flexible. However, xml.Marshal will serialize `Zs: []int{0,1}` only as `<x><y><z>0</z><z>1</z></y></x>` - see for example http://play.golang.org/p/L2DkdvaFR2. Because SecurityGroupInfo type is also used by the ec2test server to generate the DescribeSecurityGroups response, the generated XML will in fact be _invalid_ according to AWS XML schema (ipRanges>item>cidrIp,cidrIp rather than ipRanges>item>cidrIp,item>cidrIp). Go on the other hand handles both the valid and invalid XML the same way.

In summary, I realized _only_ the ec2test server needs to change how it serializes security groups and the ec2 client code can continue to use []string for SourceIPs. So I'll revert the changes around SourceIPs and I'm sorry for being stupid :)

**Comment by dimitern** [comment]
Thanks for the review @niemeyer !

I've replied to your comments and I'll propose the updated code soon. Regarding your last comments:
1. It's difficult to refer to the _exact_ time and version when some of the changes happened, as Amazon is not publishing detailed change logs for each new version. I used the official WSDL schema for EC2 when comparing the changes (see https://code.launchpad.net/~dimitern/goamz/update-aws-api-version-to-latest/+merge/243057/comments/605109). However, in addition to pointing out the actual changes to structs and fields, I'm currently going over each set of release notes (https://aws.amazon.com/releasenotes/Amazon-EC2) from 2011-12-15 to the latest 2014-07-30 to point out what changed when.
2. vpcId as an explicit argument is used by CreateSecurityGroup, CreateSubnet, VPCs (as ids []string) and DeleteVpc only. The other methods where a VPC ID can be specified, like RunInstances or CreateNetworkInterface, take a params struct with a VPCId field. None of the other existing methods care about VPC ID othewise, so I think I've covered all cases.

**Comment by dimitern** [comment]
Here is a list of AWS API versions that introduced some of the changes in this PR (where I could determine when the change happened), according to the published official [Release Notes](https://aws.amazon.com/releasenotes/Amazon-EC2):
- Instance.AssociatePublicIP - [2013-08-20](https://aws.amazon.com/releasenotes/Amazon-EC2/1428013879853538)
- Instance.EBSOptimized - [2012-07-31](https://aws.amazon.com/releasenotes/Amazon-EC2/3818672553907399)
- Instance.SRIOVNetSupport and Image.SRIOVNetSupport - 2013-10-15 - no release notes, but added to the [WSDL schema](https://s3.amazonaws.com/ec2-downloads/2013-10-15.ec2.wsdl).

**Comment by dimitern** [comment]
Since this is a blocker for us and there is no strong disagreement about the proposal, I'm going to merge this.
